src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,150 @@
+#include "event_set.h"
+
+#include <ucs/debug/memtrack.h>
+#include <ucs/debug/log.h>
+#include <ucs/debug/assert.h>
+
+#include <sys/epoll.h>
+#include <string.h>
+#include <errno.h>
+
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#define UCS_ASYNC_EPOLL_MAX_EVENTS      16
+
+ucs_status_t ucs_event_set_create(ucs_sys_event_set_t **event_set_p)
+{
+    ucs_sys_event_set_t *event_set;
+    ucs_status_t status;
+
+    event_set = ucs_malloc(sizeof(ucs_sys_event_set_t), "ucs_sys_event_set");
+    if(event_set == NULL){
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    /* Create epoll set the thread will wait on */
+    event_set->epfd = epoll_create(1);
+    if (event_set->epfd < 0) {
+        ucs_error("epoll_create() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    status = UCS_OK;
+    *event_set_p = event_set;
+    goto out_create;
+
+err_free:
+    ucs_free(event_set);
+
+err:
+
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               int events, void *udata)
+{
+    struct epoll_event event;
+    ucs_status_t status;
+    int ret;
+
+    memset(&event, 0, sizeof(event));
+    event.events = events;
+    event.data.fd = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        status = UCS_ERR_IO_ERROR;
+        goto err_add_event;
+    }
+
+    return UCS_OK;
+
+err_add_event:
+
+    return status;
+
+}
+
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+{
+    ucs_status_t status;
+    int ret;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_DEL, event_fd, NULL);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, DEL, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        status = UCS_ERR_IO_ERROR;
+        goto err_del_event;
+    }
+
+    return UCS_OK;
+
+err_del_event:
+
+    return status;
+}
+
+//
+// event 
+//  epoll:  epoll_event.data.ptr 
+//  kqueue: kevent.udata
+//
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+                                ucs_status_t (*event_set_handler)(int nready,
+                                                                  void *event))
+{
+    int nready;
+    ucs_status_t status;
+//    struct epoll_event events[UCS_ASYNC_EPOLL_MAX_EVENTS];
+    struct epoll_event *event;
```

Comment:
rename event -> events
also, we can allocate it as local var, no need to use malloc